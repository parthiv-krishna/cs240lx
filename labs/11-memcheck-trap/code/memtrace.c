// engler: cs240lx simple system to do memory tracing:
//
//  1. traps on each load or store to a range of memory.
//  2. calls a client supplied handler with all the registers
//     and the address of the fault.
//  3. client code can modify the provided registers to change
//     behavior and/or skip the faulting instruction.
// 
// makes it easy to build a variety of tools.
// 
// notes:
//   - i think pull the single step part out: that is pretty clean.
//     run_one_inst(regs, handler);
//     trap_memory(handler);
//
//   - in general our subsystems have bundled too much functionality,
//     which makes them a pain to compose.  we want to turn the code 
//     more inside out.  but then the 
//   - for pinning: we could give a way to tag an entry with a new domain.
//     you'd go in and change it in the pt and in the TLB for consistency.
#include "memtrace-internal.h"
#include "breakpoint.h"

enum { memtrace_dom = 2 };

// default: print stuff.
int memtrace_verbose_p = 1;

// total number of faults.
static volatile unsigned nfaults = 0;

// pointer to client handler to call on each domain fault.
//   - currently only have a single handler: easy to allow
//     multiple.
static memtrace_fn_t client_handler;

// pre-computed domain reg value specifying the
// domains to set for trapping heap access
static unsigned trap_heap_access;
// pre-computed domain reg value specifying the 
// domains to set for not trapping heap accesses
static unsigned no_trap_heap_access;

// remove access to <memtrace_dom> so get faults.
void memtrace_trap_enable(void) {
    staff_domain_access_ctrl_set(trap_heap_access);
}
// allow access to <memtrace_dom> so don't get faults.
void memtrace_trap_disable(void) {
    staff_domain_access_ctrl_set(no_trap_heap_access);
}

// handle the single-step mis-match exception: 
//   1. disable the mismatch breakpoint
//   2. re-enable trapping
//   3. jump back to super mode [easy to make more flexible]
//
// todo:
//   1. check that we did indeed get a mismatch and 
//      not something else.
//   2. use watchpoints to handle LDM etc.
static void 
prefetch_abort_memtrace(uint32_t *regs, uint32_t spsr, uint32_t pc) {
    if(!was_debug_fault())
        panic("was not a debug fault: %b\n", 
            prefetch_print_reason(pc));

    assert(mode_get(spsr) == USER_MODE);
    assert(regs[MCHK_CPSR_OFF] == spsr);

    mem_debug("mismatch about to resume back to: %x!\n", pc);

    // switch back to privileged mode and re-enable trapping.
    regs[MCHK_CPSR_OFF] = mode_set(spsr, SUPER_MODE);
    brkpt_disable0();
    memtrace_trap_enable();
    switchto(regs);
}

// handle a domain fault so that we can trace each load/store:
//   1. disable trapping
//   2. call client code.
//
// if client wants load/store executed, then we side-step the need
// to understand instruction semantics by running in single-step 
// mode:
//   1. set a mismatch exception on the load/store instruction
//   2. change the mode to USER (so that mismatch works)
//   3. jump back --- after the instruction completes the prefetch 
//      abort will run.
static void 
data_abort_memtrace(uint32_t *regs, uint32_t spsr, uint32_t pc) {
    uint32_t reason = data_abort_reason();
    uint32_t fault_addr = data_abort_addr();

    if(reason != DOMAIN_SECTION_FAULT)
        panic("was not a debug fault: %b\n", 
                data_abort_print_reason(pc));

    mem_debug("got a domain section fault at pc=%x: address %x\n", pc, fault_addr);
    assert(regs[MCHK_PC_OFF] == pc);
    assert(regs[MCHK_CPSR_OFF] == spsr);
        
    assert(spsr == spsr_get());
    assert(mode_get(spsr) == SUPER_MODE);

    nfaults++;
    mode_get_sp_lr(regs);
    assert(regs[MCHK_CPSR_OFF] == spsr);

    // if we have a client handler and it says to skip
    memtrace_trap_disable();
    if(client_handler(pc, fault_addr, data_fault_from_ld())) {
        mem_debug("set a mismatch on pc=%x [regs[15]=%x]\n", pc, regs[15]);
        brkpt_mismatch_set0_raw(pc);
        regs[16] = mode_set(regs[16], USER_MODE);
    }
    switchto(regs);
    not_reached();
}

void memtrace_on(void) {
    demand(client_handler, not initialized!);
    assert(mmu_is_enabled());
    assert(mode_is_super());

    // enable debugging so we can do single-stepping.
    cp14_enable();
    brkpt_disable0();

    // trap all heap access.
    memtrace_trap_enable();
}

void memtrace_off(void) {
    // shouldn't happen.
    demand(client_handler, not initialized!);
    assert(mmu_is_enabled());

    // stop trapping accesses.
    memtrace_trap_disable();
}

// in order to make sanity checking for the lab easier we 
// we force:
//   - a single fixed domain 
//   - a single fixed region
//   - must trace the entire segment.
// these restrictions are obviously ridiculous.
//
// if you have multiple checkers that check disjoint regions
// a speed hack would be to use different domains.
int memtrace_trap_region(procmap_t *pmap, void *addr, unsigned nbytes) {
    assert(nbytes == 1024*1024);

    // just for simplicity: assert we have a single domain id 
    assert(pmap->dom_ids == (1<<1));
    // [implied by above]: the memtrace_dom is avail.
    assert(!bit_isset(pmap->dom_ids, memtrace_dom));

    // get rid of the domains and go back to the old procmap.
    pr_ent_t *e = procmap_lookup(pmap, addr);
    e->dom = memtrace_dom;

    // precompute the domain register values: might be 
    // easier to just do rmw of what's in there. 
    trap_heap_access = dom_perm(pmap->dom_ids, DOM_client);
    pmap->dom_ids |= 1<<memtrace_dom;
    no_trap_heap_access = dom_perm(pmap->dom_ids, DOM_client);

    return memtrace_dom;
}

// start tracing:
//  - setup our exception handlers;
//  - change permissions so that we can access memory from user level.
//  - start pinned virtual memory.
void memtrace_init(procmap_t *pmap, memtrace_fn_t h, uint32_t vecs[]) {
    demand(!client_handler, for today: should only call once);
    client_handler = h;

    // for today's lab: make sure we always track something.
    if(!bit_isset(pmap->dom_ids, memtrace_dom))
        panic("calling memtrace_init but not tracing any region\n");

    // install our handlers [note: if <vecs> doesn't interact
    // with full_except, this won't have any effect.]
    full_except_set_prefetch(prefetch_abort_memtrace);
    full_except_set_data_abort(data_abort_memtrace);

    // start: when MMU is turned on we will start
    // trapping!
    staff_pin_mmu_on_v(pmap, vecs);

    assert(mmu_is_enabled());
}
