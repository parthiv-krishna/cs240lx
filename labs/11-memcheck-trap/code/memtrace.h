#ifndef __MEMTRACE_H__
#define __MEMTRACE_H__
#include "rpi.h"
#include "procmap.h"

// we mix the public and internal interfaces to keep stuff simple: easy to 
// seperate.

// call first.
//
// setup the mappings in <pmap>; do any other
// setup for memory tracing (e.g., shadow mem).
//
// install the handlers <vec> [this part is janky since
// we don't override anything]

// client supplied handler to call on each faulting
// load or store:
//  - the address of the faulting instruction <pc>
//  - the faulting address <addr>
//  - whether it was a load (<load_p>=1) or store (<load_p>=0)
// 
//  return value: should we resume instruction?
//      - 1 yes
//      - 0 no.
//      [this lets client emulate or whatever]
//
// TODO:
//  - currently this interface just works for clients that only need
//    read-only observations of the address stream: likely would 
//    make sense to allow mutation of the given address etc.
//    should give an additional method that takes all the registers
typedef int (*memtrace_fn_t)(uint32_t pc, uint32_t addr, unsigned load_p);

// initialize memtracing: have to have a follow on memtrace_on call
// if you want to start tracing.  
void memtrace_init(procmap_t *pmap, memtrace_fn_t h, uint32_t vec[]);

// default setup: kmalloc starts at 1MB and is 1MB big.
//
// if you want shadow memory will have to make a different
// version.

// trap on each load from or store to [addr,addr+nbytes)
//  - must call before memtrace_init
int memtrace_trap_region(procmap_t *pmap, void *addr, unsigned nbytes) ;

static inline void 
memtrace_init_default_v(memtrace_fn_t h, uint32_t vecs[]) {
    unsigned mb = 1024*1024; 
    kmalloc_init_set_start(mb, mb);
    procmap_t p = procmap_default_mk(1);
    if(memtrace_trap_region(&p, kmalloc_heap_start(), 1024*1024) < 0)
        panic("tracking failed\n");
    memtrace_init(&p, h, vecs); 
}

#include "full-except.h"

static inline void 
memtrace_init_default(memtrace_fn_t h) {
    void *v = full_except_get_vec();
    memtrace_init_default_v(h, v);
}

// turn memtracing on.
void memtrace_on(void);
// turn memtracing off.
void memtrace_off(void);

// remove r/w access by changing domain perm
void memtrace_trap_enable(void);
// enable r/w access by changing domain perm
void memtrace_trap_disable(void);
// is trapping off?
int memtrace_is_trap_off(void);

// run <fn(arg)> with memory tracing: 
//  - returns 
// result  
//  - must have done <memtrace_init>
static inline int 
memtrace_fn(int (*fn)(void *), void *arg) {
    memtrace_on();
    int ret = fn(arg);
    memtrace_off();
    return ret;
}

extern int memtrace_verbose_p;
static inline void 
memtrace_verbose_set(int verbose_p) {
    memtrace_verbose_p = verbose_p;
}
#define mem_debug(msg, args...) \
    do { if(memtrace_verbose_p) debug(msg,args); } while(0)

#endif
