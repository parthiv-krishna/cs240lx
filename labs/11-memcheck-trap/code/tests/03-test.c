// sanity check: 
//   check that we can take a domain protection fault and resume.
//
//   to isolate: we install our own exception handlers so
//   that memtrace system doesn't see the data abort.
#include "memtrace.h"
#include "mmu.h"

static volatile uint32_t expected_pc;

static int mem_handler(uint32_t pc, uint32_t addr, unsigned load_p) {
    panic("should not get a fault!\n");
}

void data_abort_vector(unsigned pc) {
    // put32/get32 these should be at the same location on everyone's pi
    trace("have a data abort at pc=%x, expected: %x\n", pc, expected_pc);
    assert(pc == expected_pc);
    trace("going to turn off tracing and resume!\n");
    memtrace_trap_disable();
}

void notmain() {
    trace("test that memtrace_init sets things up and doesn't crash\n");

    assert(!mmu_is_enabled());

    // override the memtrace exception handlers
    extern uint32_t default_vec_ints[];
    memtrace_init_default_v(mem_handler, default_vec_ints);
    memtrace_on();

    assert(mmu_is_enabled());

    // we disable trapping so, shouldn't trap.
    trace("about allocate: should not fault\n");
    memtrace_trap_disable();

    // these calls should not trap since we've disabled.
    uint32_t *p = kmalloc(4);
    trace("going to write: should succeed\n");
    put32(p, 13);
    assert(get32(p) == 13);

    // we enable trapping so should trap.
    trace("disabled heap access: should trap\n");
    memtrace_trap_enable();
    expected_pc = (uint32_t)get32;
    assert(get32(p) == 13);

    // one more trap to make sure resumption is ok.
    memtrace_trap_enable();
    trace("about to write: should get a trap\n");
    expected_pc = (uint32_t)put32;
    put32(p, 14);
    assert(get32(p) == 14);
    trace_clean_exit("success!!\n");
}
