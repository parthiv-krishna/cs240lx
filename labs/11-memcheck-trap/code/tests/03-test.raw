find-ttyusb.c:find_ttyusb:55:FOUND: </dev/ttyUSB0>
opened tty port </dev/ttyUSB0>.
pi-install: tty-usb=</dev/ttyUSB0> program=<./tests/03-test.bin> about to boot
HASH: simple_boot: sending 21960 bytes, crc32=49e5e03
waiting for a start
putting code
pi sent print: <<Flynn Dreilinger Lab 4>: success: Received the program!
>
bootloader: Done.
listening on ttyusb=</dev/ttyUSB0>
TRACE:test that memtrace_init sets things up and doesn't crash
memtrace_dom = 2
TRACE:about allocate: should not fault
TRACE:going to write: should succeed
TRACE:disabled heap access: should trap
TRACE:have a data abort at pc=0x802c, expected: 0x802c
TRACE:going to turn off tracing and resume!
TRACE:about to write: should get a trap
TRACE:have a data abort at pc=0x8024, expected: 0x8024
TRACE:going to turn off tracing and resume!
TRACE:success!!
DONE!!!

Saw done

bootloader: pi exited.  cleaning up
