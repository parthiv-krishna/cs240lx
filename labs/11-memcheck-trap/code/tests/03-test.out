TRACE:test that memtrace_init sets things up and doesn't crash
TRACE:about allocate: should not fault
TRACE:going to write: should succeed
TRACE:disabled heap access: should trap
TRACE:have a data abort at pc=0x802c, expected: 0x802c
TRACE:going to turn off tracing and resume!
TRACE:about to write: should get a trap
TRACE:have a data abort at pc=0x8024, expected: 0x8024
TRACE:going to turn off tracing and resume!
TRACE:success!!
